File: fourteen.js
function countCollatzSeqStepsIn(num) {

	var current = num
	var steps = 1
	var ended = false

	while (!ended) {
		current = returnNextCollatzStep(current)
		steps++
		if (current === 1) {
			ended = true
		}
	}
	return steps
}



function returnNextCollatzStep(num){
	if (num % 2 === 0) {
		return num / 2
	}
	else {
		return (num * 3) + 1
	}
}

var highest = 0
var collatzNum = 0

for (var i = 1; i < 1000001; i++) {
	var result = countCollatzSeqStepsIn(i)
	if (result > highest) {
		collatzNum = i
		highest = result
	}
}
console.log(collatzNum)
console.log(highest)

File: fourteen.rb
require "pry"
resolved_collatzes = {}

def sequence_length_of(num, hash)

	return hash[num] if hash[num]

	result = num
	steps = 1

	loop do
		result = calc_collatz_step(result)
		steps += 1
		if result === 1
			hash[num] = steps
			return steps
		elsif hash[result]
			steps += hash[result] - 1
			hash[num] = steps
			return steps
		end
	end
end



def calc_collatz_step(num)
	if num % 2 === 0
		return num / 2
	else
		return (num * 3) + 1
	end
end

highest = 0
i = 1
ans = 0

while i < 1000000
	res = sequence_length_of(i, resolved_collatzes)
	if res > highest
		ans = i
		highest = res
	end
	i += 1
end

puts ans